clang/include/clang/Tooling/DependencyScanning/DependencyScanningCASFilesystem.h

```diff
@@ -0,0 +1,129 @@
+//===- DependencyScanningCASFilesystem.h - clang-scan-deps fs ---*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_TOOLING_DEPENDENCYSCANNING_DEPENDENCYSCANNINGCASFILESYSTEM_H
+#define LLVM_CLANG_TOOLING_DEPENDENCYSCANNING_DEPENDENCYSCANNINGCASFILESYSTEM_H
+
+#include "clang/Basic/LLVM.h"
+#include "clang/Lex/PreprocessorExcludedConditionalDirectiveSkipMapping.h"
+#include "clang/Tooling/DependencyScanning/DependencyScanningFilesystem.h"
+#include "llvm/ADT/StringMap.h"
+#include "llvm/ADT/StringSet.h"
+#include "llvm/CAS/CASID.h"
+#include "llvm/CAS/ThreadSafeFileSystem.h"
+#include "llvm/Support/Allocator.h"
+#include "llvm/Support/ErrorOr.h"
+#include "llvm/Support/VirtualFileSystem.h"
+#include <mutex>
+
+namespace llvm {
+namespace cas {
+class CachingOnDiskFileSystem;
+} // namespace cas
+} // namespace llvm
+
+namespace clang {
+namespace tooling {
+namespace dependencies {
+
+// CAS based Dependency Scanning Woker Filesystem.
+// This should be a ThreadSafeFilesystem that you can create proxy with
+// `createThreadSafeproxyFS()`.
+class DependencyScanningCASFilesystem
+    : public DependencyScanningWorkerFilesystem {
```

Comment:
Is inheriting from `DependencyScanningWorkerFilesystem` necessary?

Reply:
This makes code in DependencyScanningWorker and DependencyScanningAction not to diverge. But on the second look, that might not be correct. Let me redo that.